docs/source/tutorials/QAOA_funcs.py

```diff
@@ -0,0 +1,439 @@
+###
+### functions for QAOA problems
+###
+
+from typing import List, Tuple, Callable, Any
+import tensorcircuit as tc
+import numpy as np
+import tensorflow as tf
+import matplotlib.pyplot as plt
+from IPython.display import clear_output
+from functools import partial
+import scipy.optimize as optimize
+
+Array = any
+Tensor = Any
+
+
+# moved
+def QUBO_to_Ising(Q: List[list]) -> Tuple[List[list], list, float]:
+    """
+    Cnvert the Q matrix into a the indication of pauli terms, the corresponding weights, and the offset.
+    The outputs are used to construct an Ising Hamiltonian for QAOA.
+
+    :param Q: The n-by-n square and symmetric Q-matrix.
+    :return pauli_terms: A list of 0/1 series, where each element represents a Pauli term.
+    A value of 1 indicates the presence of a Pauli-Z operator, while a value of 0 indicates its absence.
+    :return weights: A list of weights corresponding to each Pauli term.
+    :return offset: A float representing the offset term of the Ising Hamiltonian.
+    """
+
+    # input is n-by-n symmetric numpy array corresponding to Q-matrix
+    # output is the components of Ising Hamiltonian
+
+    n = Q.shape[0]
+
+    # square matrix check
+    if Q[0].shape[0] != n:
+        raise ValueError("Matrix is not a square matrix.")
+
+    offset = (
+        np.triu(Q, 0).sum() / 2
+    )  # Calculate the offset term of the Ising Hamiltonian
+    pauli_terms = []  # List to store the Pauli terms
+    weights = (
+        -np.sum(Q, axis=1) / 2
+    )  # Calculate the weights corresponding to each Pauli term
+
+    for i in range(n):
+        term = np.zeros(n)
+        term[i] = 1
+        pauli_terms.append(
+            term.tolist()
+        )  # Add a Pauli term corresponding to a single qubit
+
+    for i in range(n - 1):
+        for j in range(i + 1, n):
+            term = np.zeros(n)
+            term[i] = 1
+            term[j] = 1
+            pauli_terms.append(
+                term.tolist()
+            )  # Add a Pauli term corresponding to a two-qubit interaction
+
+            weight = (
+                Q[i][j] / 2
+            )  # Calculate the weight for the two-qubit interaction term
+            weights = np.concatenate(
+                (weights, weight), axis=None
+            )  # Add the weight to the weights list
+
+    return pauli_terms, weights, offset
+
+
+def Ising_loss(c: tc.Circuit, pauli_terms: List[list], weights: list) -> float:
+    """
+    computes the loss function for the Ising model based on a given quantum circuit,
```

Comment:
spelling: computes -> Computes

docstring would benefit from an explanation of what the loss function is